chimedb/dataflag/opinion/orm.py

```diff
@@ -0,0 +1,270 @@
+                % type(password).__name__
+            )
+
+        try:
+            user_row = (
+                MediaWikiUser.select(MediaWikiUser.user_id, MediaWikiUser.user_password)
+                .where(MediaWikiUser.user_name == user)
+                .get()
+            )
+        except pw.DoesNotExist:
+            raise UserWarning("User '%s not found." % user)
+        stored_hash = user_row.user_password.split(":")
+
+        # All passwords are currently stored in the following way, but MediaWiki supports different
+        # hashing algorithms (see https://www.mediawiki.org/wiki/Manual:User_table#user_password).
+        if len(stored_hash) != 4 or stored_hash[1] != "B":
+            raise ValidationError(
+                "'user_password' field for user '%s' has unknown format: %s"
+                % (user, stored_hash)
+            )
+        _, _, salt, stored_hash = stored_hash
+
+        salted_hash = salt + "-" + hashlib.md5(password.encode("utf-8")).hexdigest()
+        hashed_salt = hashlib.md5(salted_hash.encode("utf-8")).hexdigest()
+        if hashed_salt != stored_hash:
+            raise UserWarning("Wrong username or password.")
+        return user, user_row.user_id
+
+
+class DataFlagOpinion(DataFlag):
+    """A persons opinion on flagging a range of data.
+
+    Attributes
+    ----------
+    type : DataFlagOpinionType
+        The type of flag.
+    user : MediaWikiUser
+        The user who made this opinion.
+    decision : str
+        Decision about the daa: "good", "bad" or "unsure".
+    start_time, finish_time : double
+        The start and end times as UNIX times.
+    metadata : dict
+        A JSON object with extended metadata. See below for guidelines.
+    creation_time, last_edit : double
+        The time of creation and last edit as UNIX times.
+
+    Notes
+    -----
+    To ensure that the added metadata is easily parseable, it should adhere
+    to a rough schema. The following common fields may be present:
+
+    `instrument` : optional
+        The name of the instrument that the flag opinion applies to. If not set,
+        assumed to apply to all instruments.
+    `freq` : optional
+        A list of integer frequency IDs that the flag opinion applies to. If not
+        present the flagging opinion is assumed to apply to *all* frequencies.
+    `inputs` : optional
+        A list of integer feed IDs (in cylinder order) that the flagging opinion applies
+        to. If not present the flagging opinion is assumed to apply to *all* inputs. For
+        this to make sense an `instrument` field is also required.
+
+    Any other useful metadata can be put straight into the metadata field,
+    though it must be accessed directly.
+    """
+
+    type = pw.ForeignKeyField(DataFlagOpinionType, backref="opinions")
+    user = pw.ForeignKeyField(MediaWikiUser, backref="opinions")
+    choices_decision = ["good", "bad", "unsure"]
```

Comment:
Can't you use an enum here? I think we have our own peewee type implementation for it.

Reply:
Yes: `from chimedb.core import EnumField`.  Downgrades itself to a varchar when necessary.